src/universe/visualization/img_tools.py

```diff
@@ -14,3 +19,352 @@ def rotate(img: np.ndarray, angle: float) -> np.ndarray:
+    def draw_on(self, image: np.ndarray) -> np.ndarray:
+        ...
+
+
+class LinesLayer(_Layer):
+    """
+    A layer with lines that can be drawn on an image.
+    """
+
+    def __init__(self, lines: List[Line2d], color: Tuple[int, int, int], endpoints: bool = False,
+                 line_width: float = 0.004, opacity: float = 0.75, name: Optional[str] = None):
+        """
+        Creates a new layer with the given lines, color, opacity and name.
+
+        :param lines: lines to draw
+        :param color: color of the lines (RGB)
+        :param endpoints: whether to draw circles at the endpoints of the lines
+        :param line_width: this value multiplied by the width of the image is the line width in pixels
+        :param opacity: opacity of the lines (0-1)
+        :param name: name of the layer
+        """
+        if not lines:
+            print('No lines to draw' if name is None else f'No lines to draw on layer {name}')
+        super().__init__(color, opacity, name)
+        self.lines = lines
+        self.endpoints = endpoints
+        self.line_width = line_width
+
+    def draw_on(self, image: np.ndarray) -> np.ndarray:
+        """
+        Draws the lines on the given image.
+
+        :param image: image to draw the lines on
+        :return: image with the lines drawn on it
+        """
+        return draw_lines(image, self.lines, self.color, self.endpoints, self.line_width, self._opacity)
+
+
+class MaskLayer(_Layer):
+    """
+    A layer with a mask that can be drawn on an image.
+    """
+
+    def __init__(self, mask: np.ndarray, color: Tuple[int, int, int], threshold: int = 30,
+                 opacity: float = 0.75, name: Optional[str] = None):
+        """
+        Creates a new layer with the given mask, color, opacity and name.
+
+        :param mask: mask to draw
+        :param color: color of the mask (RGB)
+        :param opacity: opacity of the mask (0-1)
+        :param threshold: threshold of the mask (0-255)
+        :param name: name of the layer
+        """
+        if mask is None:
+            print('No mask to draw' if name is None else f'No mask to draw on layer {name}')
+        super().__init__(color, opacity, name)
+        self.mask = mask
+        self.threshold = threshold
+
+    def draw_on(self, image: np.ndarray) -> np.ndarray:
+        """
+        Draws the mask on the given image.
+
+        :param image: image to draw the mask on
+        :return: image with the mask drawn on it
+        """
+        return draw_mask(image, self.mask, self.color, self.threshold, self._opacity)
+
+
+class LayeredImage:
+    """
+    An image that can have layers drawn on it.
+    """
+
+    def __init__(self, image: np.ndarray):
+        """
+        Creates a new layered image with the given image. The original image is not modified.
+
+        :param image: image to draw the layers on
+        """
+        if image is None:
+            raise ValueError('Image does not exist')
```

Comment:
also here

Reply:
I believe it IS a problem if there is no image to draw on.

Reply:
but WHEN ?

Reply:
I don't understand. You want to create a LayeredImage object without an image?